File: 1-二叉树/1-二叉树/Akagi201/binary-tree.c

/*
 * @file binary-tree.c
 *  
 * @author Akagi201 
 * @date 2014/02/16 
 *  
 * binary tree C file 
 *
 */

#include "binary-tree.h"
#include <stdio.h>
#include <stdlib.h>

/*
 * @brief input pre-order -> make binary tree
 *  
 * pre-order, space 
 *  
 * @param[in] tree: binary tree root node 
 * @param[out] null 
 * @return null 
 * @sa 
 * @note 
 * @warning 
 */
void make_tree(struct btnode **tree){
    char element;
    scanf("%c", &element);

    if ((' ' == element)) {
        *tree = NULL;
    }else{
        if ((*tree = (struct btnode *)malloc(sizeof(struct btnode))) == NULL) {
            printf("make_tree malloc failed.\n");
            exit(1);
        }

        (*tree)->data = element;

        printf("make_tree succeed with element:%c.\n", element);
        make_tree(&((*tree)->left));
        make_tree(&((*tree)->right));
    }

    return;
}

void pre_order(struct btnode *tree){
    if (tree != NULL) {
        printf("visited: %c.\n", tree->data);
        pre_order(tree->left);
        pre_order(tree->right);
    }

    return;
}

void in_order(struct btnode *tree){
    if (tree != NULL) {
        in_order(tree->left);
        printf("visited: %c.\n", tree->data);
        in_order(tree->right);
    }

    return;
}

void post_order(struct btnode *tree){
    if (tree != NULL) {
        post_order(tree->left);
        post_order(tree->right);
        printf("visited: %c.\n", tree->data);
    }

    return;
}

void level_order(struct btnode *tree){

}

int main(void){
    struct btnode *tree;
    make_tree(&tree);
    pre_order(tree);
    in_order(tree);
    post_order(tree);
    level_order(tree);
    return 0;
}

File: 1-树/0-二叉树/徐嘉轶/binary-tree.js
function TreeNode(value){
    this.value = value;
    return this;
}

TreeNode.prototype = {
    preOrder : function(cb){
        // abcdefg
        cb(this);
        this.left && this.left.preOrder(cb);
        this.right && this.right.preOrder(cb);
    },
    inOrder: function(cb){
        // cbdafeg
        this.left && this.left.inOrder(cb);
        cb(this);
        this.right && this.right.inOrder(cb);
    },
    postOrder: function(cb){
        this.left && this.left.postOrder(cb);
        this.right && this.right.postOrder(cb);
        cb(this);
    },
    levelOrder: function(cb){
        var tree = this;
        var current = null;
        var list = [tree];
        while(list.length){
            var len = list.length;
            current = list[0];
            current.left && list.push(current.left);
            current.right && list.push(current.right);
            cb(list.shift());
        }
    }
}


function handler(tree){
    process.stdout.write(tree.value);
}


var tree = new TreeNode("a");
tree.left = new TreeNode("b");
tree.left.left = new TreeNode("c");
tree.left.right = new TreeNode("d");
tree.right = new TreeNode("e");
tree.right.left = new TreeNode("f");
tree.right.right = new TreeNode("g");


tree.preOrder(handler);
console.log();
tree.inOrder(handler);
console.log();
tree.postOrder(handler);
console.log();
tree.levelOrder(handler);